iocage_lib/ioc_upgrade.py

```diff
@@ -140,44 +146,26 @@ def upgrade_jail(self):
                         callback=self.callback
                     )
             else:
-                try:
-                    iocage_lib.ioc_exec.InteractiveExec(
-                        fetch_cmd,
-                        self.path.replace('/root', ''),
-                        uuid=self.uuid,
-                        unjailed=True
-                    )
-                except iocage_lib.ioc_exceptions.CommandFailed:
-                    self.__rollback_jail__()
-                    msg = f'Upgrade failed! Rolling back jail'
+                iocage_lib.ioc_exec.InteractiveExec(
+                    fetch_cmd,
+                    self.path.replace('/root', ''),
+                    uuid=self.uuid,
+                    unjailed=True
+                )
+
+            if not os.path.islink(self.freebsd_install_link):
+                    msg = f'Upgrade failed, nothing to install after fetch!'
                     iocage_lib.ioc_common.logit(
                         {
-                            "level": "EXCEPTION",
-                            "message": msg
+                            'level': 'EXCEPTION',
+                            'message': msg
                         },
                         _callback=self.callback,
                         silent=self.silent
                     )
 
-            if not self.interactive:
-                while not self.__upgrade_install__(tmp.name):
-                    pass
-            else:
-                # FreeBSD update loops 3 times
-                for _ in range(3):
-                    try:
-                        self.__upgrade_install__(tmp.name)
-                    except iocage_lib.ioc_exceptions.CommandFailed:
-                        self.__rollback_jail__()
-                        msg = f'Upgrade failed! Rolling back jail'
-                        iocage_lib.ioc_common.logit(
-                            {
-                                'level': 'EXCEPTION',
-                                'message': msg
-                            },
-                            _callback=self.callback,
-                            silent=self.silent
-                        )
+            while os.path.islink(self.freebsd_install_link):
+                self.__upgrade_install__(tmp.name)
```

Comment:
Is it possible that this never ends ?

Reply:
No, there is no code path in freebsd-update install that would exit on 0 and leave ${BDHASH}-install behind.

Reply:
> There is, but only after `freebsd-update upgrade` when it wants you to reboot after it installs a new kernel / before it deletes old shared libraries.

Touché - I clearly wasn't precise enough, sorry. What I meant to say is, that freebsd-update install either moves the installation forward or exits != 0. It doesn't exit on 0 and gets stuck in whatever it was doing beforehand. You can trust its exit code - it won't say "ok" if there is a problem.

The logic is basically:
- Check params - if there's a problem, exit 1 (unless ${BDHASH}-install doesn't exist and fetch was called, in which case it exits 0 - but this code path in unreachable in our case)
- install_files:
  + If !kernel_done: install kernel, exit 1 on error. touch kernel_done; exit 0
  + install other files - exit 1 on any error
  + rename ${BDHASH}-install to ${BDHASH}-rollback

Reply:
Quite silly of me, as actually the whole point of this patch is to look if ${BDHASH}-install was left behind while `freebsd-update install` exited on 0 to know that additional invocations are required. ^_^

The point is, that, unless there are bugs introduced in future versions of freebsd-update, this loop should be safe by all means.

Reply:
@sonicaj So the correct answer should've been:

"No, there is no code path in `freebsd-update install` that would exit on 0 and leave ${BDHASH}-install behind _without progressing the installation process_."

@cperciva Does that sound acceptable?

Reply:
@grembo @cperciva thank you for clarifying. @skarekrow are you comfortable with this loop ? I am not sure how to proceed with this, if anything changes on freebsd-update end even as a bug for some time, this is going to repeat itself endlessly. If you think we are good, i am okay with it then.

Reply:
@skarekrow i feel like maybe putting an upper limit here ? i.e 5 maybe ? not sure, i just don't want us to be affected by any unwanted upstream changes. @grembo would you have any leads ?

Reply:
@sonicaj If it helps: If FreeBSD-update changed in the future, this would also affect ezjail and iocell (and probably many other handcrafted solutions). The advantage iocage has over ezjail and iocell with this patch is that it can tell apart „finished“ and „something broke“ (they run “while ok”, while this code runs “while file is there, expect ok”). Besides, a looping upgrade process will certainly cause an issue being opened *immediately*.

In general, I’m all for defensive coding and my first instinct was to put in a runaway handler - just looking the at remainder of iocage I think it’s written to trust return codes from the various tools used.

That said, if these concerns delay the patch, I’m more than happy to add an upper limit + exception - I’m absolutely not religious about it. I would select a fairly high number like 50 iterations in this case though and fail the upgrade process once it’s reached.

Reply:
I think it's reasonable for us to rely on exit codes of our upstream projects. Which we do in a lot of cases. I'm not against defensive programming either, but like @grembo mentioned, a user will very likely notice that the script is indeed looping and kill it, and (hopefully) promptly filing an issue.

What I want to avoid is us having to low of a ceiling in case upstream changes how things are done (perhaps it becomes a wrapper to pkg base in the future -- who knows), so I agree with @grembo's suggestion of 50. That seems rather high, but it's high enough we don't worry about prematurely killing things, and low enough that the program won't spin for ages.

Regardless, this essentially replaces what we did before which was a fairly naive loop that just hoped for the best. I don't think this PR will bring us any additional risk of undesired behavior, and the fix is worth the risk imo.

So tldr; let's add a 50 iteration ceiling, that's a good enough compromise.

Reply:
@grembo can you please add a ceiling of 50 iterations ? Thank you
@skarekrow i agree with the compromise ;)

Reply:
@sonicaj I implemented a runaway check, also fixed an indentation I got wrong in the previous version and corrected a function description I stumbled upon (__upgrade_install__ doesn't return anything any more and even before it didn't return the exit code, but True/False/None).